float32 distance
float32 angle
bool valid

